---
title: "Mapping Illumina reads with Bowtie for use with CNVer"
author: "Simon Renny-Byfield"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

This document details my attempts at running bowtie to generate suitable `.bam` files for input into `CNVer`. Analysis of reads will be done using [paap](https://github.com/RILAB/paap), which records quality, trims reads, throws poor reads out and maps reads using (at the moment BWA-MEM). I can utilize the trimmed quality output of paap for input into bowtie, so as not to have to do the qulaity check and trim twice. I'll write my own script to run bowtie. But first, in order to get bowtie up and running I will need to generate a bowtie index for the reference genome.

##Building a bowtie index##

I used this the followign command to generate an index for the masked version of the APGv_3 referecnce sequence.

```
#!/bin/bash
#SBATCH -D /home/sbyfield/CNVer
#SBATCH -o /home/sbyfield/CNVer/logs/out_log-%j.txt
#SBATCH -e /home/sbyfield/CNVer/logs/err_log-%j.txt
#SBATCH -J index

date

cmd="bowtie-build Zmays_284_AGPv3.hardmasked_shortName.fa AGPv3"

echo $cmd

eval$cmd

date
```

Next you have to map reads to the reference genome using bowtie. This can be achieved with a command something like this:

```
#!/bin/bash
#SBATCH -D /home/sbyfield/CNVer
#SBATCH -o /home/sbyfield/CNVer/logs/out_log-%j.txt
#SBATCH -e /home/sbyfield/CNVer/logs/err_log-%j.txt
#SBATCH -J bowtie

echo "Starting Job: "
date

cmd="gunzip test_data/TIL01_3510_3807_3510_N_TIP521_4_R1.fastq.gz"
echo $cmd
eval $cmd

cmd="bowtie AGPv3 test_data/TIL01_3510_3807_3510_N_TIP521_4_R1.fastq -p 12 -S -v 3 -a -m 600 --best --strata | samtools view -b -S -o bams/R1.bam -"
echo $cmd
eval $cmd

cmd="gunzip test_data/TIL01_3510_3807_3510_N_TIP521_4_R2.fastq.gz"
echo $cmd
eval $cmd

cmd="bowtie AGPv3 test_data/TIL01_3510_3807_3510_N_TIP521_4_R2.fastq -p 12 -S -v 3 -a -m 600 --best --strata | samtools view -b -S -o bams/R2.bam -"
echo $cmd
eval $cmd

module load gcc/4.5 cnver

cmd="interleave_bam bams/R1.bam bams/R2.bam bams/interleaved.bam"
echo $cmd
eval $cmd

echo "Ending Job: "
date
```

I found this [website](http://moo.nac.uci.edu/~hjm/cnver_bduc_howto.html#_step_4_convert_sort_bowtie_output_to_be_compatible_with_cnver) with a description of how to prepare the bam files for input into cnver. I will follow their advice and use the same parameters to see if that makes a differnce. **It didn't solve the problem**. This issue is documented in another markdown script [issues.Rmd](https://github.com/XLEvolutionist/CNVer/blob/master/issues.Rmd) and I have resorted to re-doing the mapping using the above example as a template.

The latest scrip to use bowtie is as follows.

```
#!/bin/bash
#SBATCH -D /home/sbyfield/CNVer/test_data
#SBATCH -o /home/sbyfield/CNVer/logs/out_log-%j.txt
#SBATCH -e /home/sbyfield/CNVer/logs/err_log-%j.txt
#SBATCH -J bowtie
#SBATCH --cpus-per-task 10
#SBATCH --array=0-19
#SBATCH --mem=10000

declare -a list
readarray list < ../paired.fastq.txt

files=${list[$SLURM_ARRAY_TASK_ID]}
echo $files

file1=`echo $files | cut -d " " -f1`
file2=`echo $files | cut -d " " -f2`

echo "Starting Job on $file: "
date

cmd="mkfifo my_queue.$file1.fifo"
echo $cmd
eval $cmd

cmd="mkfifo my_queue.$file2.fifo"
echo $cmd
eval $cmd

echo $file1
echo $file2
zcat $file1 >  my_queue.$file1.fifo & 
zcat $file2 >  my_queue.$file2.fifo &

bowtie ../AGPv3 -1 my_queue.$file1.fifo -2 my_queue.$file2.fifo -p 8 -S -v 2 -a -m 600 --best --strata | samtools view -b -S - | samtools sort -n -m 8GB - - > ../bams/$file1.combine

rm my_queue.$file1.fifo
rm my_queue.$file2.fifo

echo "Ending Job: "
date
```