<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="author" content="Simon Renny-Byfield" />

<meta name="date" content="2014-12-19" />

<title>Trial Run with CNVer</title>




<link href="data:text/css,body%20%7B%0A%20%20background%2Dcolor%3A%20%23fff%3B%0A%20%20margin%3A%201em%20auto%3B%0A%20%20max%2Dwidth%3A%20700px%3B%0A%20%20overflow%3A%20visible%3B%0A%20%20padding%2Dleft%3A%202em%3B%0A%20%20padding%2Dright%3A%202em%3B%0A%20%20font%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0A%20%20font%2Dsize%3A%2014px%3B%0A%20%20line%2Dheight%3A%201%2E35%3B%0A%7D%0A%0A%23header%20%7B%0A%20%20text%2Dalign%3A%20center%3B%0A%7D%0A%0A%23TOC%20%7B%0A%20%20clear%3A%20both%3B%0A%20%20margin%3A%200%200%2010px%2010px%3B%0A%20%20padding%3A%204px%3B%0A%20%20width%3A%20400px%3B%0A%20%20border%3A%201px%20solid%20%23CCCCCC%3B%0A%20%20border%2Dradius%3A%205px%3B%0A%0A%20%20background%2Dcolor%3A%20%23f6f6f6%3B%0A%20%20font%2Dsize%3A%2013px%3B%0A%20%20line%2Dheight%3A%201%2E3%3B%0A%7D%0A%20%20%23TOC%20%2Etoctitle%20%7B%0A%20%20%20%20font%2Dweight%3A%20bold%3B%0A%20%20%20%20font%2Dsize%3A%2015px%3B%0A%20%20%20%20margin%2Dleft%3A%205px%3B%0A%20%20%7D%0A%0A%20%20%23TOC%20ul%20%7B%0A%20%20%20%20padding%2Dleft%3A%2040px%3B%0A%20%20%20%20margin%2Dleft%3A%20%2D1%2E5em%3B%0A%20%20%20%20margin%2Dtop%3A%205px%3B%0A%20%20%20%20margin%2Dbottom%3A%205px%3B%0A%20%20%7D%0A%20%20%23TOC%20ul%20ul%20%7B%0A%20%20%20%20margin%2Dleft%3A%20%2D2em%3B%0A%20%20%7D%0A%20%20%23TOC%20li%20%7B%0A%20%20%20%20line%2Dheight%3A%2016px%3B%0A%20%20%7D%0A%0Atable%20%7B%0A%20%20margin%3A%201em%20auto%3B%0A%20%20border%2Dwidth%3A%201px%3B%0A%20%20border%2Dcolor%3A%20%23DDDDDD%3B%0A%20%20border%2Dstyle%3A%20outset%3B%0A%20%20border%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0A%20%20border%2Dwidth%3A%202px%3B%0A%20%20padding%3A%205px%3B%0A%20%20border%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0A%20%20border%2Dwidth%3A%201px%3B%0A%20%20border%2Dstyle%3A%20inset%3B%0A%20%20line%2Dheight%3A%2018px%3B%0A%20%20padding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0A%20%20border%2Dleft%2Dstyle%3A%20none%3B%0A%20%20border%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0A%20%20background%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0A%0Ap%20%7B%0A%20%20margin%3A%200%2E5em%200%3B%0A%7D%0A%0Ablockquote%20%7B%0A%20%20background%2Dcolor%3A%20%23f6f6f6%3B%0A%20%20padding%3A%200%2E25em%200%2E75em%3B%0A%7D%0A%0Ahr%20%7B%0A%20%20border%2Dstyle%3A%20solid%3B%0A%20%20border%3A%20none%3B%0A%20%20border%2Dtop%3A%201px%20solid%20%23777%3B%0A%20%20margin%3A%2028px%200%3B%0A%7D%0A%0Adl%20%7B%0A%20%20margin%2Dleft%3A%200%3B%0A%7D%0A%20%20dl%20dd%20%7B%0A%20%20%20%20margin%2Dbottom%3A%2013px%3B%0A%20%20%20%20margin%2Dleft%3A%2013px%3B%0A%20%20%7D%0A%20%20dl%20dt%20%7B%0A%20%20%20%20font%2Dweight%3A%20bold%3B%0A%20%20%7D%0A%0Aul%20%7B%0A%20%20margin%2Dtop%3A%200%3B%0A%7D%0A%20%20ul%20li%20%7B%0A%20%20%20%20list%2Dstyle%3A%20circle%20outside%3B%0A%20%20%7D%0A%20%20ul%20ul%20%7B%0A%20%20%20%20margin%2Dbottom%3A%200%3B%0A%20%20%7D%0A%0Apre%2C%20code%20%7B%0A%20%20background%2Dcolor%3A%20%23f7f7f7%3B%0A%20%20border%2Dradius%3A%203px%3B%0A%20%20color%3A%20%23333%3B%0A%7D%0Apre%20%7B%0A%20%20white%2Dspace%3A%20pre%2Dwrap%3B%20%20%20%20%2F%2A%20Wrap%20long%20lines%20%2A%2F%0A%20%20border%2Dradius%3A%203px%3B%0A%20%20margin%3A%205px%200px%2010px%200px%3B%0A%20%20padding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0A%20%20background%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0A%0Acode%20%7B%0A%20%20font%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0A%20%20font%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0A%20%20padding%3A%202px%200px%3B%0A%7D%0A%0Adiv%2Efigure%20%7B%0A%20%20text%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0A%20%20background%2Dcolor%3A%20%23FFFFFF%3B%0A%20%20padding%3A%202px%3B%0A%20%20border%3A%201px%20solid%20%23DDDDDD%3B%0A%20%20border%2Dradius%3A%203px%3B%0A%20%20border%3A%201px%20solid%20%23CCCCCC%3B%0A%20%20margin%3A%200%205px%3B%0A%7D%0A%0Ah1%20%7B%0A%20%20margin%2Dtop%3A%200%3B%0A%20%20font%2Dsize%3A%2035px%3B%0A%20%20line%2Dheight%3A%2040px%3B%0A%7D%0A%0Ah2%20%7B%0A%20%20border%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0A%20%20padding%2Dtop%3A%2010px%3B%0A%20%20padding%2Dbottom%3A%202px%3B%0A%20%20font%2Dsize%3A%20145%25%3B%0A%7D%0A%0Ah3%20%7B%0A%20%20border%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0A%20%20padding%2Dtop%3A%2010px%3B%0A%20%20font%2Dsize%3A%20120%25%3B%0A%7D%0A%0Ah4%20%7B%0A%20%20border%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0A%20%20margin%2Dleft%3A%208px%3B%0A%20%20font%2Dsize%3A%20105%25%3B%0A%7D%0A%0Ah5%2C%20h6%20%7B%0A%20%20border%2Dbottom%3A%201px%20solid%20%23ccc%3B%0A%20%20font%2Dsize%3A%20105%25%3B%0A%7D%0A%0Aa%20%7B%0A%20%20color%3A%20%230033dd%3B%0A%20%20text%2Ddecoration%3A%20none%3B%0A%7D%0A%20%20a%3Ahover%20%7B%0A%20%20%20%20color%3A%20%236666ff%3B%20%7D%0A%20%20a%3Avisited%20%7B%0A%20%20%20%20color%3A%20%23800080%3B%20%7D%0A%20%20a%3Avisited%3Ahover%20%7B%0A%20%20%20%20color%3A%20%23BB00BB%3B%20%7D%0A%20%20a%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0A%20%20%20%20text%2Ddecoration%3A%20underline%3B%20%7D%0A%20%20a%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0A%20%20%20%20text%2Ddecoration%3A%20underline%3B%20%7D%0A%0A%2F%2A%20Class%20described%20in%20https%3A%2F%2Fbenjeffrey%2Ecom%2Fposts%2Fpandoc%2Dsyntax%2Dhighlighting%2Dcss%0A%20%20%20Colours%20from%20https%3A%2F%2Fgist%2Egithub%2Ecom%2Frobsimmons%2F1172277%20%2A%2F%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%2F%2A%20Keyword%20%2A%2F%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%2F%2A%20DataType%20%2A%2F%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%2F%2A%20DecVal%20%28decimal%20values%29%20%2A%2F%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20BaseN%20%2A%2F%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20Float%20%2A%2F%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20Char%20%2A%2F%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20String%20%2A%2F%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%2F%2A%20Comment%20%2A%2F%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%2F%2A%20OtherToken%20%2A%2F%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%2F%2A%20AlertToken%20%2A%2F%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%2F%2A%20Function%20calls%20%2A%2F%20%0Acode%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%2F%2A%20ErrorTok%20%2A%2F%0A%0A" rel="stylesheet" type="text/css" />

</head>

<body>



<div id="header">
<h1 class="title">Trial Run with CNVer</h1>
<h4 class="author"><em>Simon Renny-Byfield</em></h4>
<h4 class="date"><em>2014-12-19</em></h4>
</div>


<p>This document tracks my first attempt at running <a href="http://compbio.cs.toronto.edu/CNVer/">CNVer</a> on a test data set. CNVer is a program that estimates presence absence variants and copy number variants using mate pair/paired end Illumina sequence data. The purpose of this excercise is to identfiy PAV and CNV in a wild population of Teosinte (Palmar Chico) and in turn estimate the impact of PAVs/CNVs on estimates of the <em>SFS</em>, <em>pi</em> and <em>Tajimaâ€™s D</em>.</p>
<p>As the authors of <code>CNVer</code> suggest I will use <a href="http://bowtie-bio.sourceforge.net/index.shtml">bowtie</a> to map reads to the AGPv3 reference genome. A detailed description of how this mapping is achieved is available <a href="https://github.com/XLEvolutionist/CNVer/blob/master/bowtie_mapping.Rmd">here</a>. The mapping is done like this:</p>
<pre><code>#!/bin/bash
#SBATCH -D /home/sbyfield/CNVer
#SBATCH -o /home/sbyfield/CNVer/logs/out_log-%j.txt
#SBATCH -e /home/sbyfield/CNVer/logs/err_log-%j.txt
#SBATCH -J bowtie

echo &quot;Starting Job: &quot;
date

cmd=&quot;gunzip test_data/TIL01_3510_3807_3510_N_TIP521_4_R1.fastq.gz&quot;
echo $cmd
eval $cmd

cmd=&quot;bowtie AGPv3 test_data/TIL01_3510_3807_3510_N_TIP521_4_R1.fastq -p 12 -S -v 3 -a -m 600 --best --strata | samtools view -b -S -o bams/R1.bam -&quot;
echo $cmd
eval $cmd

cmd=&quot;gunzip test_data/TIL01_3510_3807_3510_N_TIP521_4_R2.fastq.gz&quot;
echo $cmd
eval $cmd

cmd=&quot;bowtie AGPv3 test_data/TIL01_3510_3807_3510_N_TIP521_4_R2.fastq -p 12 -S -v 3 -a -m 600 --best --strata | samtools view -b -S -o bams/R2.bam -&quot;
echo $cmd
eval $cmd

module load gcc/4.5 cnver

cmd=&quot;interleave_bam bams/R1.bam bams/R2.bam bams/interleaved.bam&quot;
echo $cmd
eval $cmd

echo &quot;Ending Job: &quot;
date</code></pre>
<p>Next It is important to estimate the mean and SD of insert size in your mapping data so that CNVer can do a good job of identifying discordant mate pairs. This can be achieved using a program distributed with CNVer called <code>find_mean_sd</code> and the command look likes this:</p>
<pre><code>#!/bin/bash -l
#SBATCH -D /home/sbyfield/CNVer
#SBATCH -J MeanL
#SBATCH -o /home/sbyfield/CNVer/out-%j.txt
#SBATCH -e /home/sbyfield/CNVer/error-%j.txt
#SBATCH --mem=20000

module load gcc/4.5 cnver

echo &quot;Starting Jog: &quot;
date

cmd=&quot;find_mean_sd bams/interleaved.bam bam 50 1000 75&quot;
echo $cmd
eval $cmd

echo &quot;End Job: &quot;
date</code></pre>
<p>arguments are &lt;mapping file&gt; &lt;format&gt; &lt;min length&gt; &lt;max length&gt; &lt;seq length&gt;.</p>
<div id="running-cnver-with-a-test-dataset-of-one-individual-til01" class="section level2">
<h2>Running CNVer with a test dataset of one individual (TIL01)</h2>
<p>Once the mean and standard deviation of insert size has been determined you can then run CNVer to estimate CNVs and PNVs. The command looks something like this:</p>
<pre><code>#!/bin/bash -l
#SBATCH -D /home/sbyfield/CNVer
#SBATCH -J CNVER
#SBATCH -o /home/sbyfield/CNVer/logs/out-%j.txt
#SBATCH -e /home/sbyfield/CNVer/logs/error-%j.txt
#SBATCH --mem=24000

echo &quot;Starting CNVer Job: &quot;
date
echo &quot;Loading module CNVer&quot;
module load gcc/4.5 cnver

cmd=&quot;cnver.pl --map_list ~/CNVer/test.file.list.txt --ref_folder ~/CNVer/refZea/ --work_dir ~/CNVer/output --read_len 75 --mean_insert 111 --stdev_insert 25 --min_mps 4 --ref_names refChr.txt&quot;
echo $cmd
eval $cmd

echo &quot;Ending Job: &quot;
date</code></pre>
<p>A list, and details of the arguments are provided below:</p>
<ul>
<li><strong>â€“map_list</strong> A list of the bam files to be used for mapping. In the original case only one file.</li>
<li><strong>â€“ref_folder</strong> The reference companion package, created as described in the <a href="https://github.com/XLEvolutionist/CNVer/blob/master/companionPackage.Rmd">companion package</a> script, and is found in <code>/home/srbyfield/CNVer/refZea/</code>.</li>
<li><strong>â€“work_dir</strong> Where you would like the output to be stored.</li>
<li><strong>â€“read_len 75</strong> The read length of the Illumina data</li>
<li><strong>â€“mean_insert</strong> The mean insert size. Can be calculated as described above using the <code>find_mean_sd</code> script provided with the <code>CNVer</code> package.</li>
<li><strong>â€“stdev_insert</strong> The SD of the insert size. Also calculated with the <code>find_mean_sd</code> script provided with the <code>CNVer</code> package.</li>
<li><strong>â€“min_mps</strong> Min number of reads required to make a cluster.</li>
<li><strong>â€“ref_names</strong> This is a file which details the chromosomes of interest (i.e.Â in this case only chr 10).</li>
</ul>
<p>I originaly run the above call with <code>--min_mps</code> set to 10, but this prodiced a graph with no nodes or edges. So i adjusted this parameter to be 4. Keep in mind that the <code>--min_mps</code> needs to be at least 2.</p>
<p>I am having issues getting the pipeline to run. The issues are documented in <a href="https://github.com/XLEvolutionist/CNVer/blob/master/issues.Rmd">issues.Rmd</a>, and a modifed bowtie mapping run has been added to <a href="https://github.com/XLEvolutionist/CNVer/blob/master/bowtie_mapping.Rmd">bowtie mapping</a></p>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
