---
title: "CNVer error, Walk-through"
author: "Simon Renny-Byfield"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

This is a simple document attempting to detail the issue with CNVer, and follow the steps I have taken to try to resolve the problem. The problem is actually that the final output .cvs file is empty (as are many of the intermediate files). The `CNVer` run is performed like this:

```
#!/bin/bash -l
#SBATCH -D /home/sbyfield/CNVer
#SBATCH -J CNVER
#SBATCH -o /home/sbyfield/CNVer/logs/out-%j.txt
#SBATCH -e /home/sbyfield/CNVer/logs/error-%j.txt
#SBATCH --mem=24000

echo "Starting CNVer Job: "
date
echo "Loading module CNVer"
module load gcc/4.5 cnver

cmd="cnver.pl --map_list ~/CNVer/test.file.list.txt --ref_folder ~/CNVer/refZea/ --work_dir ~/CNVer/output --read_len 100 --mean_insert 112 --stdev_insert 23 --min_mps 4 --ref_names ~/CNVer/refChr.txt"
echo $cmd
eval $cmd

echo "Ending Job: "
date
```

Scrolling through the log file I see this output:
```
Starting CNVer Job:
Thu Dec  4 19:22:53 PST 2014
Loading module CNVer
cnver.pl --map_list ~/CNVer/test.file.list.txt --ref_folder ~/CNVer/refZea/ --work_dir ~/CNVer/output --read_len 100 --mean_insert 112 --stdev_insert 23 --min_mps 4 --ref_names ~/CNVer/refChr.txt
Exec at 19:22:53, Thu Dec 4, 2014 (/share/apps/cnver-0.8.1/src/maps2bin /home/sbyfield/CNVer/test.file.list.txt /home/sbyfield/CNVer/refZea//allchr.txt bam &);  /share/apps/cnver-0.8.1/src/cluster/concordancy_analysis  -r100 -d10000000 -l43 -u181 -m/home/sbyfield/CNVer/test.file.list.txt -n/home/sbyfield/CNVer/refZea//allchr.txt -ibam ...

Exec at 0:43:34, Fri Dec 5, 2014 /share/apps/cnver-0.8.1/src/make_simple_coverage /home/sbyfield/CNVer/output/mapping_files/10.rmap /home/sbyfield/CNVer/output/10//10.scov 10 149632204...

Exec at 0:43:51, Fri Dec 5, 2014 /share/apps/cnver-0.8.1/src/find_spikes /home/sbyfield/CNVer/output/10//10.scov  0 15 | grep -v low | awk '{ s=$2; if (s>1) {s=s-2} else {s=0} print $1,s,$3+2,$4 }' > /home/sbyfield/CNVer/output/10//10.spikes...

Exec at 0:44:3, Fri Dec 5, 2014 /share/apps/cnver-0.8.1/src/make_ns /home/sbyfield/CNVer/refZea//fasta_files_folder/10.fa 10 > /home/sbyfield/CNVer/output/10//10.ns...

Exec at 0:44:5, Fri Dec 5, 2014 cat /home/sbyfield/CNVer/refZea//contig_breaks_folder/10.cbs /home/sbyfield/CNVer/refZea//repeat_regions_folder/10.rep /home/sbyfield/CNVer/output/10//10.ns /home/sbyfield/CNVer/output/10//10.spikes | sort -k 2n | /share/apps/cnver-0.8.1/src/interval_join > /home/sbyfield/CNVer/output/10//10.masks...

Exec at 0:44:8, Fri Dec 5, 2014 /share/apps/cnver-0.8.1/src/make_gc_map /home/sbyfield/CNVer/refZea//fasta_files_folder/10.fa /home/sbyfield/CNVer/output/10//10.masks /home/sbyfield/CNVer/output/10//10.scov /home/sbyfield/CNVer/output/10//10.gc 10 500 50...

Exec at 0:44:31, Fri Dec 5, 2014 awk '{ if ($6 == 0) print $0 }'  /home/sbyfield/CNVer/output/mapping_files/10.mmap  | sort -k3n,3 -k4n,4 -k1n,1  | /share/apps/cnver-0.8.1/src/cluster/cluster_matepairs --mean=112 --stdev=23 --colID=-1 --colDist=0 --colChr=1 --colLeft=2 --colRight=3 --colTemplate=4 --baseLenFactor=1 --mdJoinTolerance=6 --type=0 --concise=1 > /home/sbyfield/CNVer/output/10//10.t0...

Exec at 0:44:32, Fri Dec 5, 2014 awk '{ if ($6 == 1) print $0 }'  /home/sbyfield/CNVer/output/mapping_files/10.mmap  | sort -k4n,4 -k3n,3 -k1n,1  | /share/apps/cnver-0.8.1/src/cluster/cluster_matepairs --mean=112 --stdev=23 --colID=-1 --colDist=0 --colChr=1 --colLeft=2 --colRight=3 --colTemplate=4 --baseLenFactor=1 --mdJoinTolerance=6 --type=1 --concise=1 > /home/sbyfield/CNVer/output/10//10.t1...

Exec at 0:44:33, Fri Dec 5, 2014 awk '{ if ($6 == 2) print $0 }'  /home/sbyfield/CNVer/output/mapping_files/10.mmap  | sort -k3n,3 -k4n,4 -k1n,1  | /share/apps/cnver-0.8.1/src/cluster/cluster_matepairs --mean=112 --stdev=23 --colID=-1 --colDist=0 --colChr=1 --colLeft=2 --colRight=3 --colTemplate=4 --baseLenFactor=1 --mdJoinTolerance=6 --type=2 --concise=1 > /home/sbyfield/CNVer/output/10//10.t2...

Exec at 0:44:34, Fri Dec 5, 2014 awk '{ if ($6 == 3) print $0 }'  /home/sbyfield/CNVer/output/mapping_files/10.mmap  | sort -k4n,4 -k3n,3 -k1n,1  | /share/apps/cnver-0.8.1/src/cluster/cluster_matepairs --mean=112 --stdev=23 --colID=-1 --colDist=0 --colChr=1 --colLeft=2 --colRight=3 --colTemplate=4 --baseLenFactor=1 --mdJoinTolerance=6 --type=3 --concise=1 > /home/sbyfield/CNVer/output/10//10.t3...

Exec at 0:44:36, Fri Dec 5, 2014 grep -h EDGE /home/sbyfield/CNVer/output/10//10.t[0123] | /share/apps/cnver-0.8.1/src/cluster/screen_contig_breaks --mean=112 --stdev=23 --tolerance=181 --breaksFile = /home/sbyfield/CNVer/refZea//contig_breaks_folder/10.cbs | awk '{ if (($5 >= 4) && ($8 < 10000000)) print $0 }' > /home/sbyfield/CNVer/output/10//10.links...

Exec at 0:44:36, Fri Dec 5, 2014 /share/apps/cnver-0.8.1/src/axt_to_binary /home/sbyfield/CNVer/refZea//self_alignments_folder/10.axt 149632204 /home/sbyfield/CNVer/output/10//10.edges...
Axtbin
Exec at 0:45:19, Fri Dec 5, 2014 /share/apps/cnver-0.8.1/src/make_reference_graph /home/sbyfield/CNVer/output/10//10.edges /home/sbyfield/CNVer/output/10//10.graph...

Exec at 0:45:19, Fri Dec 5, 2014 /share/apps/cnver-0.8.1/src/fill_reference_graph /home/sbyfield/CNVer/output/10//10.graph 149632204...

Exec at 0:45:19, Fri Dec 5, 2014 /share/apps/cnver-0.8.1/src/make_donor_graph /home/sbyfield/CNVer/output/10//10.graph /home/sbyfield/CNVer/output/10//10.links /home/sbyfield/CNVer/output/10//10.scov /home/sbyfield/CNVer/output/10//10.masks /home/sbyfield/CNVer/output/10//10.gc 1> /home/sbyfield/CNVer/output/10//10.problem 2> /home/sbyfield/CNVer/output/10//10.graphinfo...

Exec at 0:45:25, Fri Dec 5, 2014 perl /share/apps/cnver-0.8.1/src/flow_solve/make_cs2_graph.pl   10 10.problem 2 0 20 1 0 /share/apps/cnver-0.8.1/src/cs2-4.6/...
Usage: /share/apps/cnver-0.8.1/src/flow_solve/make_cs2_graph.pl genome_length num_reads maxlines infile baseflow freelimit allownone cs2_dir
Exec at 0:45:25, Fri Dec 5, 2014 /share/apps/cnver-0.8.1/src/report_cnvs /home/sbyfield/CNVer/output/10//10.graphinfo /home/sbyfield/CNVer/output/10//10.problem.out 100 0 2> /home/sbyfield/CNVer/output/10//10.cnvs.raw 1> /home/sbyfield/CNVer/output/10//10.used_dgs...

Exec at 0:45:26, Fri Dec 5, 2014 cat /home/sbyfield/CNVer/output/10//10.cnvs.raw | /share/apps/cnver-0.8.1/src/intRemove2  /home/sbyfield/CNVer/refZea//contig_breaks_folder/10.cbs > /home/sbyfield/CNVer/output/10//10.cnvs.raw.screened...

Exec at 0:45:26, Fri Dec 5, 2014 /share/apps/cnver-0.8.1/src/smoother.sh 5 1000 10.cnvs.raw.screened > /home/sbyfield/CNVer/output/10//10.cnvs.smoothed...

Exec at 0:45:26, Fri Dec 5, 2014 cat /home/sbyfield/CNVer/output/10//10.cnvs.smoothed | awk -v OFS="\t" '{ if ($4 > 0) print $1, $2, $3, "gain"; else print $1, $2, $3, "loss" }' | /share/apps/cnver-0.8.1/src/post_analysis/doc_walker /home/sbyfield/CNVer/output/10//10.scov /home/sbyfield/CNVer/output/10//10.gc | awk -v OFS="\t" '{ print $1, $2, $3, $4, $6 }' > /home/sbyfield/CNVer/output/calls//10.cnv$

Ending Job:
Fri Dec  5 00:45:26 PST 2014
```

The important thing to note is that one of the calls seems to fail due to too few command line arguments, and this call is given below. Importantly several command line arguments are missing and are listed below. I will try to address these in list order.
```
Exec at 0:45:25, Fri Dec 5, 2014 perl /share/apps/cnver-0.8.1/src/flow_solve/make_cs2_graph.pl   10 10.problem 2 0 20 1 0 /share/apps/cnver-0.8.1/src/cs2-4.6/...
Usage: /share/apps/cnver-0.8.1/src/flow_solve/make_cs2_graph.pl genome_length num_reads maxlines infile baseflow freelimit allownone cs2_dir
```
##The *genome_length* argument is missing##

Below is shown a section of the script `cnver.pl`, which controls the flow of the program, and this genome_length variable is defined in this section. 
```
  #make the donor graph
	execCommand("$CNVER_FOLDER/src/make_donor_graph $work_fol/$ref.graph $work_fol/$ref.links $work_fol/$ref.scov $work_fol/$ref.masks $work_fol/$ref.gc 1> $work_fol/$ref.problem 2> $work_fol/$ref.graphinfo");

	my $contiglen = `tail -n 1 $work_fol/$ref.graphinfo | awk '{print \$2 }' `;
	my $numreads  = `tail -n 1 $work_fol/$ref.graphinfo | awk '{print \$3 }' `;
	chomp $contiglen;
	chomp $numreads;

	chdir ($work_fol);
	#prepare graph for cs2 solver and solve it
	execCommand ("perl $CNVER_FOLDER/src/flow_solve/make_cs2_graph.pl $contiglen $numreads 10 $ref.problem 2 0 20 1 0 $CNVER_FOLDER/src/cs2-4.6/");
```
The important part is this ` my $contiglen = ``tail -n 1 $work_fol/$ref.graphinfo | awk '{print \$2 }' ``;`. So this call reads in `$ref.graphinfo` and determines the contig length from this file. Currently the file `10.graphinfo` is empty, so the call fails. So the next question is what call makes the file `10.graphinfo`... It is actually the previous call:
```
#make the donor graph
  execCommand("$CNVER_FOLDER/src/make_donor_graph $work_fol/$ref.graph $work_fol/$ref.links $work_fol/$ref.scov $work_fol/$ref.masks $work_fol/$ref.gc 1> $work_fol/$ref.problem 2> $work_fol/$ref.graphinfo");
```

The actual call used by CNVer is listed in the log file and is 
```
/share/apps/cnver-0.8.1/src/make_donor_graph /home/sbyfield/CNVer/output/10//10.graph /home/sbyfield/CNVer/output/10//10.links /home/sbyfield/CNVer/output/10//10.scov /home/sbyfield/CNVer/output/10//10.masks /home/sbyfield/CNVer/output/10//10.gc 1> /home/sbyfield/CNVer/output/10//10.problem 2> /home/sbyfield/CNVer/output/10//10.graphinfo...
```
The arguments to this call are files. They exist and are not emtpy but the first argument, the file `10.graph`, looks like this:
```
^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^A^@^@^@^@^@^@^@�4�^H^@^@^@^@+^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@
```
So what call creates this file...? The following command seems to make the `.graph` file.
```
execCommand ("$CNVER_FOLDER/src/make_reference_graph $work_fol/$ref.edges $work_fol/$ref.graph");
```
The `$ref.edges` file is empty, so no wonder the `.graph` file is empty. So back track again. What makes the `.edges` file. The program takes the `.axt` file and input. This file looks like this:
```
# lastz.v1.02.00 --hspthresh=5000 --format=axt 
#
# hsp_threshold      = 5000
# gapped_threshold   = 5000
# x_drop             = 910
# y_drop             = 9400
# gap_open_penalty   = 400
# gap_extend_penalty = 30
#        A    C    G    T
#   A   91 -114  -31 -123
#   C -114  100 -125  -31
#   G  -31 -125  100 -114
#   T -123  -31 -114   91
0 Chr1 4550 4683 10 116888684 116888805 + 7788
GTTTTAGTAACCTTGCCATGTTACTAAAACGATTAAGGGGGTGTTTGGTTACCCCTGCTAAAGTTTAGCCCGGGTCACATCAAGCGTTTGACTTTCAAATAGGAGTATGAAATATAGACCCAACCAACTGGACT
GTTCTAGCGATGCTGTCAT------------ACTGAGGGGGTGTTTGGTTCTGTGGACTAAAGTTTAGCCAGGGTCACATCAAGCGTTTGACTTTCAAGTAGGAGTATGAAATATAGACCCAACCAACTGTACT

1 Chr1 4552 4658 10 5766072 5766178 + 6153
TTTAGTAACCTTGCCATGTTACTAAAACGATTAAGGGGGTGTTTGGTTACCCCTGCTAAAGTTTAGCCCGGGTCACATCAAGCGTTTGACTTTCAAATAGGAGTATG
TTCAGTGTTCTTGTCTCCTTTTGGACCCTACTCAGGCTGTATTTGGTTGCTCCTGCTAAAGTTTAGCTCGGGTCACATCAAGCGTTTGACTTTTAAATAGGAGTATG
```
Notice that the *hit* chromosome is labelled like this `Chr1` but it should be just `1`. This is probably the source of the problem. I need to use unix commands in order to remove the `Chr` component of the name. Something like this should do the trick:

```
#/usr/bin/bash

files=$(ls *.axt)

for i in ${files}; do
  sed -i -e 's/Chr//g' $i
done
```

I tested this step of the analysis by running the same call as described in the log file to see if removing Chr from the `.axt` file helped. The command I used is as follows:
```
#!/bin/bash
/share/apps/cnver-0.8.1/src/axt_to_binary /home/sbyfield/CNVer/refZea//self_alignments_folder/10.axt 149632204 /home/sbyfield/CNVer/output/10//10.edges
```
and this produces output! and the `.edges` file is being made!

The file is made, but there seems to another issue. The call fails midway through making the file with the following error:
```
Exec at 1:23:20, Sun Dec 7, 2014 /share/apps/cnver-0.8.1/src/make_reference_graph /home/sbyfield/CNVer/output/10//10.edges /home/sbyfield/CNVer/output/10//10.graph...
Merging the same edge???
```

##Not all reads in the TIL01 and TIL11 data have pairs##

On further inspection I discovered that the two fastq files containing the paired reads do not contain exactly equivalent sets of reads. That is some pairs are missing from either one of the files. I big problem for CNVer and this violates one of the [assumptions](http://compbio.cs.toronto.edu/cnver/README-0.8.1.txt) of `CNVer`. This is not a problem for the first 20 teosinte parents that we already have, but is a problem with the TIL01 and TIL11 samples that we need to quality check the CNV calls. In response I wrote a script in perl to filter the HUGE fastq files using an index and a fastq stream from Bio::Index::Fastq and Bio::SeqIO::fastq. Here is the script:
```
#!/usr/bin/perl
use strict;
use warnings;
use Bio::Index::Fastq;
use Bio::SeqIO::fastq;

#Simom Renny-Byfield, UC Davis, Dec `14

#this code takes paired fastq files and filters the reads for only those that have 
#a match in both files. That is, any read that has had it's mate discarded will be thrown
#out

#usage <script.pl> <read1> <read2>

#make an index for one of the fastq files.
my $Index_File_Name0 = "$ARGV[0].idx";
#make the index for the first file
my $inx0 = Bio::Index::Fastq->new(
        '-filename' => $Index_File_Name0,
        '-write_flag' => 1);
    $inx0->make_index($ARGV[0]);

#now set up a fastq stream with Bio::SeqIO:fastq
my $in = Bio::SeqIO->new(-format    => 'fastq-sanger',
                           -file      => $ARGV[1]);
#now create two output streams, one for each of the pair                           
my $out0 = Bio::SeqIO->new(-format    => 'fastq-sanger',
                            -file      => ">$ARGV[0].clean");   
my $out1 = Bio::SeqIO->new(-format    => 'fastq-sanger',
                            -file      => ">$ARGV[1].clean");

#now open the Bio::SeqIO::fastq stream (the second file) and try to grab the sequence in
#the first file. If not present skip to the next sequence. This essentially abandons non-
#paired reads.
  
# $seq is a Bio::SeqIO::fastq object
while (my $seq1 = $in->next_seq) {
      #do stuff
      my $id1 = $seq1->id;
      if ( my $seq0 = $inx0->fetch($id1) ) {
        $out0->write_seq($seq0);
      	$out1->write_seq($seq1);
      }#if
      else {
      	print "Did not match, next seq\n";
      }#else
}#while    

exit;
```
Essentailly this script looks for read IDs from the fastq stream in the index of the opposing file. Any reads without matches are discarded. I imagine this will be pretty slow (`BioPerl` is not the quickest of cats), and we may need a different approach when we come to quality check the paired reads for the rest of the data. In fact it takes an age just to generate the `.idx` file (currently at 5 hours), so we will definitely need to not generate the situation where pairs are lost. I has now been running for 16 hrs... far too slow. It hasn't even finished indexing the fastq file.

An alternative approach presents itself: After some research I discovered that the program [sickle](http://compbio.cs.toronto.edu/cnver/README-0.8.1.txt) will trim and maintain the paired end information in paired fastq files. This is probably the way forward in terms of trimming.

I subsequently went back and used the merged (unfiltered at this point), reads that apparently have the same number of reads, and mapped these using bowtie. Then `interleaved_bam` script and sorted the reads based on name using `samtools sort -n`. 

I then used `samtools flagstat <in.bam>` to check that the `.bam` contains properly paired reads, what is needed for `CNVer`. [Here](https://www.biostars.org/p/12475/) is an interesting blog post about the meaning of the output of `samtools flagstat`. An example:
```
20968800 + 0 mapped (100.00%:nan%)
20968800 + 0 paired in sequencing
```
These two numbers should be the same if the reads are "paired in sequencing". This does not produce the expected results:
```
992777935 + 0 in total (QC-passed reads + QC-failed reads)
0 + 0 duplicates
208214994 + 0 mapped (20.97%:-nan%)
0 + 0 paired in sequencing
0 + 0 read1
0 + 0 read2
0 + 0 properly paired (-nan%:-nan%)
0 + 0 with itself and mate mapped
0 + 0 singletons (-nan%:-nan%)
0 + 0 with mate mapped to a different chr
0 + 0 with mate mapped to a different chr (mapQ>=5)
```
`samtools` does not recognize the reads as paired, put at least there is an even number of reads. Odd would indicate that there are some reads missing.

